bus-mapping/src/evm/opcodes/extcodecopy.rs

```diff
@@ -56,90 +60,190 @@ fn gen_extcodecopy_step(
+        dest_offset,
+    )?;
+    state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(2), offset)?;
+    state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(3), length)?;
+
+    for (field, value) in [
+        (CallContextField::TxId, U256::from(state.tx_ctx.id())),
+        (
+            CallContextField::RwCounterEndOfReversion,
+            U256::from(state.call()?.rw_counter_end_of_reversion as u64),
+        ),
+        (
+            CallContextField::IsPersistent,
+            U256::from(state.call()?.is_persistent as u64),
+        ),
+    ] {
+        state.call_context_read(&mut exec_step, state.call()?.call_id, field, value);
+    }
+
+    let is_warm = state.sdb.check_account_in_access_list(&external_address);
     state.push_op_reversible(
         &mut exec_step,
         RW::WRITE,
         TxAccessListAccountOp {
             tx_id: state.tx_ctx.id(),
-            address: address.to_address(),
+            address: external_address,
             is_warm: true,
             is_warm_prev: is_warm,
         },
     )?;
 
-    // stack reads
-    state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(0), address)?;
-    state.stack_read(
+    let &Account { code_hash, .. } = state.sdb.get_account(&external_address).1;
+    state.account_read(
         &mut exec_step,
-        geth_step.stack.nth_last_filled(1),
-        dest_offset,
+        external_address,
+        AccountField::CodeHash,
+        code_hash.to_word(),
+        code_hash.to_word(),
     )?;
-    state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(2), offset)?;
-    state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(3), length)?;
     Ok(exec_step)
 }
 
+fn gen_copy_steps(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+    src_addr: u64,
+    dst_addr: u64,
+    src_addr_end: u64,
+    bytes_left: u64,
+    code: Vec<u8>,
+) -> Result<Vec<(u8, bool)>, Error> {
+    let mut copy_steps = Vec::with_capacity(bytes_left as usize);
+    for idx in 0..bytes_left {
+        let addr = src_addr + idx;
+        let value = if addr < src_addr_end {
+            code[addr as usize]
+        } else {
+            0
+        };
+        copy_steps.push((value, false));
+        state.memory_write(exec_step, (dst_addr + idx).into(), value)?;
+    }
+
+    Ok(copy_steps)
+}
+
+fn gen_copy_event(
+    state: &mut CircuitInputStateRef,
+    geth_step: &GethExecStep,
+) -> Result<CopyEvent, Error> {
+    let rw_counter_start = state.block_ctx.rwc;
+    let external_address = geth_step.stack.nth_last(0)?.to_address();
+    let memory_offset = geth_step.stack.nth_last(1)?.as_u64();
+    let data_offset = geth_step.stack.nth_last(2)?.as_u64();
+    let length = geth_step.stack.nth_last(3)?.as_u64();
+
+    let mut exec_step = gen_extcodecopy_step(state, geth_step)?;
+    let code_hash = state.sdb.get_account(&external_address).1.code_hash;
+    let code = state.code(code_hash)?;
```

Comment:
You can get the `eth_types::Bytecode` here:
```suggestion
    let code: Bytecode = state.code(code_hash)?.into();
```

Reply:
Fixed in [23aaa95](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/720/commits/23aaa95f3d673b00942f76535a92177cda08a4ff).